src/haz3lcore/dynamics/Evaluator.re

```diff
@@ -130,7 +140,8 @@ let rec matches = (dp: DHPat.t, d: DHExp.t): match_result =>
   | (Inj(_, _), Cast(d, Sum(_, _), Unknown(_))) => matches(dp, d)
   | (Inj(_, _), Cast(d, Unknown(_), Sum(_, _))) => matches(dp, d)
   | (Inj(_, _), _) => DoesNotMatch
-  | (Pair(dp1, dp2), Pair(d1, d2)) =>
+  | (Pair(dp1, dp2), Pair(d1, d2))
+  | (Ap(dp1, dp2), Ap(d1, d2)) =>
```

Comment:
we probably want to special case Ap with a Tag immediately in the dp1 position as distinct from one with a var or other pattern there, which is typically not allowed (matching on arbitrary tags)